R/ic.R

```diff
@@ -5,66 +5,54 @@
 #' @return If `x` is an expression, returns the result of evaluating `x`. If `x` is missing nothing
 #'   is returned.
 #'
-#' @export
-#'
 #' @examples
 #' f <- function(x) x < 0
 #'
 #' ic(f(1))
 #'
 #' ic(f(-1))
+#'
+#' @importFrom rlang enquo quo_is_missing trace_back quo_get_expr caller_fn caller_env expr_deparse eval_tidy fn_env env_label maybe_missing
+#' @importFrom glue glue
+#' @export
 ic <- function(x) {
   # Capture the input to allow us to work with the expression and value separately
-  q <- rlang::enquo(x)
+  q <- enquo(x)
 
   # The behaviour of the function changes depending on whether input is provided or not.
-  missing_input <- rlang::quo_is_missing(q)
+  missing_input <- quo_is_missing(q)
 
   # In the event that icecream is totally disabled we will just return the input.
-  #
-  # TODO: This triggers evaluation of the input. Is this a problem?
-  if (isFALSE(getOption("icecream.enabled"))) {
-    if (!missing_input) {
-      return(x)
+  if (getOption("icecream.enabled")) {
+    trace <- trace_back()
+    num_calls <- length(trace$calls)
+
+    parent_ref <-  if (num_calls > 1) trace$calls[[num_calls - 1]][[1]] else NULL
+    ref <- attr(trace$calls[[num_calls]], "srcref")
+    loc <- src_loc(ref)
+
+    # Case when location of file is unavailable
+    if (nchar(loc) == 0) {
+      # Probs want to look at environments
+      caller <- caller_fn()
+      caller_env <- if (is.null(caller)) caller_env() else fn_env(caller)
+
+      loc <- env_label(caller_env)
+      loc <- glue("<env: {loc}>")
     }
-    return(invisible())
-  }
 
-  trace <- rlang::trace_back()
-  num_calls <- length(trace$calls)
-  parent_ref <-  if (num_calls > 1) trace$calls[[num_calls - 1]][[1]] else NULL
-  ref <- attr(trace$calls[[num_calls]], "srcref")
-  loc <- src_loc(ref)
-  if (nchar(loc) == 0) {
-    # Probs want to look at environments
-    caller_fn <- sys.function(-1)
-    if (is.null(caller_fn)) {
-      e <- sys.frame(-1)
+    # If we have inputs then we want the expression and value to be included in the context
+    # object as well.
+    if (!missing_input) {
+      deparsed_expression <- expr_deparse(quo_get_expr(q))
+      x <- eval_tidy(q)
+      ic_print(loc, parent_ref, deparsed_expression, x)
```

Comment:
Sorry, just thought: couldn't we simplify this a bit to:

``` r
  if (!missing_input) {
    deparsed_expression <- expr_deparse(quo_get_expr(q))
    x <- eval_tidy(q)
    ic_print(loc, parent_ref, deparsed_expression, x)
    return(invisible(x))
  }
  # If we get this far we know there was no input so we just print
  ic_print(loc, parent_ref)
}
# End of function
```

That means we can remove the final `if (missing_input) invisible() else invisible(x)` line.

Reply:
Yes, you're right. Such change should solve problem of invisibility in case of disabled ic. I'll commit it. Then I think we can just merge PR

Reply:
Well, `ic_print` right now return its output invisibly as a result (I also wonder if that should be the case... ). The last line should be return of just `invisible()`. But you're right, checking the condition twice is pointless. I just like when the function has only one line of return.